<a href="https://colab.research.google.com/github/haiibarose/DataSciDataEng/blob/main/code/Titanic_DataPrep.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment: Preparing Data for Analysis (Modified Titanic)

![](https://github.com/kaopanboonyuen/2110446_DataScience_2021s2/raw/main/%20files/hw.png)

In [155]:
#Import the libraries
import numpy as np
import pandas as pd

In [156]:
df = pd.read_csv('https://github.com/kaopanboonyuen/2110446_DataScience_2021s2/raw/main/datasets/hw/titanic_training_dataset_v2.csv')
df2 = df.copy()

In [157]:
df2.head(40)

PassengerId  Survived  Pclass  \
0             1       0.0     3.0   
1             2       1.0     1.0   
2             3       1.0     3.0   
3             4       1.0     1.0   
4             5       0.0     3.0   
5             6       0.0     3.0   
6             7       0.0     1.0   
7             8       0.0     3.0   
8             9       1.0     3.0   
9            10       1.0     2.0   
10           11       1.0     3.0   
11           12       1.0     1.0   
12           13       0.0     3.0   
13           14       0.0     3.0   
14           15       0.0     3.0   
15           16       1.0     2.0   
16           17       0.0     3.0   
17           18       1.0     2.0   
18           19       0.0     3.0   
19           20       1.0     3.0   
20           21       0.0     2.0   
21           22       1.0     2.0   
22           23       1.0     3.0   
23           24       1.0     1.0   
24           25       0.0     3.0   
25           26       1.0     3.0   
26           27       0.0     3.0   
27           28       0.0     1.0   
28           29       1.0     3.0   
29           30       0.0     3.0   
30           31       0.0     1.0   
31           32       1.0     1.0   
32           33       1.0     3.0   
33           34       0.0     2.0   
34           35       0.0     1.0   
35           36       0.0     1.0   
36           37       1.0     3.0   
37           38       0.0     3.0   
38           39       0.0     3.0   
39           40       1.0     3.0   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0    1.0   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                              Heikkinen, Miss. Laina  female  26.0    0.0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                            Allen, Mr. William Henry    male  35.0    0.0   
5                                    Moran, Mr. James    male   NaN    0.0   
6                             McCarthy, Mr. Timothy J    male  54.0    0.0   
7                      Palsson, Master. Gosta Leonard    male   2.0    3.0   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0    0.0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0    1.0   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0    1.0   
11                           Bonnell, Miss. Elizabeth  female  58.0    0.0   
12                     Saundercock, Mr. William Henry    male  20.0    0.0   
13                        Andersson, Mr. Anders Johan    male  39.0    1.0   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0    0.0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0    0.0   
16                               Rice, Master. Eugene    male   2.0    4.0   
17                       Williams, Mr. Charles Eugene    male   NaN    0.0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0    1.0   
19                            Masselmani, Mrs. Fatima  female   NaN    0.0   
20                               Fynney, Mr. Joseph J    male  35.0    0.0   
21                              Beesley, Mr. Lawrence    male  34.0    0.0   
22                        McGowan, Miss. Anna "Annie"  female  15.0    0.0   
23                       Sloper, Mr. William Thompson    male  28.0    0.0   
24                      Palsson, Miss. Torborg Danira  female   8.0    3.0   
25  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0    1.0   
26                            Emir, Mr. Farred Chehab    male   NaN    0.0   
27                     Fortune, Mr. Charles Alexander    male  19.0    3.0   
28                      O'Dwyer, Miss. Ellen "Nellie"  female   NaN    0.0   
29                                Todoroff, Mr. Lalio    male   NaN    0.0   
30                           Uruchurtu, Don. Manuel E    male  40.0 

In [158]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     865 non-null    float64
 2   Pclass       828 non-null    float64
 3   Name         866 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        858 non-null    float64
 7   Parch        891 non-null    int64  
 8   Ticket       852 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     803 non-null    object 
dtypes: float64(5), int64(2), object(5)
memory usage: 83.7+ KB


# 1) Load data & review the data

<font color='orange'>Q1) What is the shape of this dataset? (rows & columns)</font>

<font color='lightblue'>Answer: 891 rows & 12 columns</font>

In [159]:
df2.shape

(891, 12)

# 2) Drop unqualified variables

*   Drop variables with missing > 50%
*   Drop categorical variables with flat values > 70% (variables with the same value in the same column)

<font color='orange'>Q2) How many columns do we have left?</font>

<font color='lightblue'>Answer: 10 columns</font>





In [160]:
all_cols = df2.columns
drop_cols = list()
for col in df2:
    na = df2[col].isnull().sum()
    count = df2[col].shape[0]
    percent_miss = na*100/count
    if percent_miss > 50:
        drop_cols.append(col)
        continue
    if col in df2.select_dtypes(include=['object']):
        check_flat = df2[col].value_counts()
        percent_flat = max(check_flat)*100/sum(check_flat)
        if percent_flat > 70:
            drop_cols.append(col)

In [161]:
df2 = df2.drop(drop_cols, axis=1)

In [162]:
df2.shape
# df2.columns

(891, 10)

In [163]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     865 non-null    float64
 2   Pclass       828 non-null    float64
 3   Name         866 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        858 non-null    float64
 7   Parch        891 non-null    int64  
 8   Ticket       852 non-null    object 
 9   Fare         891 non-null    float64
dtypes: float64(5), int64(2), object(3)
memory usage: 69.7+ KB


# 3) Remove all rows with missing target (the variable "Survived")

<font color='orange'>Q3) How many rows do we have left?</font>

<font color='lightblue'>Answer: 865 rows</font>

In [164]:
df2 = df2.dropna(subset=['Survived'])
df2 = df2.reset_index()

In [165]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 865 entries, 0 to 864
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   index        865 non-null    int64  
 1   PassengerId  865 non-null    int64  
 2   Survived     865 non-null    float64
 3   Pclass       802 non-null    float64
 4   Name         840 non-null    object 
 5   Sex          865 non-null    object 
 6   Age          693 non-null    float64
 7   SibSp        832 non-null    float64
 8   Parch        865 non-null    int64  
 9   Ticket       826 non-null    object 
 10  Fare         865 non-null    float64
dtypes: float64(5), int64(3), object(3)
memory usage: 74.5+ KB


# 4) Handle outliers

For the variable “Fare”, replace outlier values with the boundary values


*   If value < (Q1 - 1.5IQR), relace with (Q1 - 1.5IQR)
*   If value > (Q3 + 1.5IQR), relace with (Q3 + 1.5IQR)

<font color='orange'>Q4) What is the mean of “Fare” after replacing the outliers (round 2 decimal points)?</font>

<font color='lightblue'>Answer: mean of 'Fare' is 24.04</font>




In [166]:
q3, q1 = np.percentile(df2['Fare'], [75 ,25])
iqr = q3 - q1
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

df2['Fare'] = np.where(df2['Fare'] < lower_bound, lower_bound, df2['Fare'])
df2['Fare'] = np.where(df2['Fare'] > upper_bound, upper_bound, df2['Fare'])

In [167]:
round(np.mean(df2['Fare']), 2)

24.04

# 5) Impute missing value



*   Show the number of missing values in each variable
*   Impute missing values with mean & mode
*   Show the number of missing values again after missing value imputation

<font color='orange'>Q5) Which variable has the largest number of missing values?</font>

<font color='lightblue'>Answer: 'Age' has largest number of missing values</font>




In [168]:
# df2.info()

In [169]:
df2.isnull().sum()

index            0
PassengerId      0
Survived         0
Pclass          63
Name            25
Sex              0
Age            172
SibSp           33
Parch            0
Ticket          39
Fare             0
dtype: int64

In [170]:
# Pclass => category => mode
df2['Pclass'].fillna(df2['Pclass'].mode().iloc[0], inplace=True)
# Name => category => mode
df2['Name'].fillna(df2['Name'].mode().iloc[0], inplace=True)
# Age => numerical => mean
df2['Age'].fillna(df2['Age'].mean(), inplace=True)
# SibSp => category => mode
df2['SibSp'].fillna(df2['SibSp'].mode().iloc[0], inplace=True)
# Ticket => category => mode
df2['Ticket'].fillna(df2['Ticket'].mode().iloc[0], inplace=True)

In [171]:
df2.isnull().sum()

index          0
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
dtype: int64

# 6) Convert categorical to numeric values



*   Drop the variables “Name” & “Ticket”
*   For the variables “Sex” & “Embarked”, perform the dummy coding and drop the first level. Also, drop those original variables (“Sex” & “Embarked”)

<font color='orange'>Q6) How many columns do we have?</font>

<font color='lightblue'>Answer: 9 columns</font>




In [172]:
df2 = df2.drop(['Name', 'Ticket'], axis=1)

In [173]:
nominal_cols = ['Sex']
dummy_df = pd.get_dummies(df2[nominal_cols], drop_first=True)
df2_with_dummy_df = pd.concat([df2, dummy_df], axis=1)
df2_with_dummy_df = df2_with_dummy_df.drop(nominal_cols, axis=1)
print(df2_with_dummy_df.shape)

(865, 9)


# 7) Partition data



*   Split train/test split with stratification using 70%:30% and random seed with 12345
*   Show a proportion between survived (1) and died (0) in all data sets (total data, train, test)

<font color='orange'>Q7) What is a proportion between survived (1) and died (0) in the training data?</font>

<font color='lightblue'>Answer: proportion between survived(1) and died(0) is $\frac{\text{0.616185}}{\text{0.383815}}$</font>



In [174]:
from sklearn.model_selection import train_test_split

y = df2_with_dummy_df['Survived']
X = df2_with_dummy_df.drop(['Survived'], axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.3, random_state=12345)

In [175]:
y.value_counts() / y.shape[0]

0.0    0.616185
1.0    0.383815
Name: Survived, dtype: float64

In [176]:
print(X_train.shape)

(605, 8)


In [177]:
print(X_test.shape)

(260, 8)


In [178]:
y_train.value_counts() / y_train.shape[0]

0.0    0.616529
1.0    0.383471
Name: Survived, dtype: float64

In [179]:
y_test.value_counts() / y_test.shape[0]

0.0    0.615385
1.0    0.384615
Name: Survived, dtype: float64